### ISSUE TITLE
feat: Hastebin route backwards compatibility 

### ISSUE THREAD

**Issue by jackiedorland** [lukewhrit/spacebin#25]
For the long-term, it would be a good idea to have a sub-route (like /haste) that would provide backward compatibility with Hastebin's endpoints.

**Comment by lukewhrit** [comment]
The thing is hastebin didn't really have endpoints. It was also GET, or POST to /. Maybe we could still do that though.

**Comment by jackiedorland** [comment]
Yeah but the point of it was to make sure that people using already-existing hastebin clients won't need to change endpoints or clients

**Comment by lukewhrit** [comment]
> Yeah but the point of it was to make sure that people using already-existing hastebin clients won't need to change endpoints or clients

Even with a sub-route you'd still need to modify a haste client. Although for most clients this is as easy as setting an environment variable.

**Comment by lukewhrit** [comment]
This is currently in the `v0.1.0` backlog but will be implemented before `v1.0.0`.

**Comment by lukewhrit** [comment]
Decided to name this endpoint `/api/v0` instead of `/haste` so that it fits with other route prefixes.

**Comment by lukewhrit** [comment]
> TLDR: Implementing the `POST` requests would be somewhat annoying due to `multipart/form-data`. Implementing the `GET` requests would be very easy.

Haste has two options for uploading:

1. Using `multipart/form-data`
2. Using `plain/text`

We already have support for `multipart/form-data`.

### `GET` requests

We already implement a nearly identical (in endpoint structure) API to `haste-server`, however where we differ is response.

This is the response body of a `GET` request to haste:

```json
{
	"data": "...",
	"key: "..."
}
``` 

This is ours:

```json
{
	"content": "...",
	"id": "..."
}
```

Implementing backwards compatibility would be very easy to implement for `GET` requests. We'd need to setup a new route to deal with the modified response body.

I still think it may not be worth it to implement this feature at the moment. It perhaps could be postponed until `Alpha v0.1.0`.

**Comment by lukewhrit** [comment]
I should've mentioned in my past comment. Haste's API is almost entirely undocumented outside of [this](https://github.com/seejohnrun/haste-server/wiki/POST-api). This makes it much harder to make an accurate recreation of their API.

By "accurate" I mean an API that mirrors request endpoints and response bodies exactly, down to the error message.

**Comment by lukewhrit** [comment]
There aren't a lot of pre-existing haste clients anyways from what I can tell. I don't think supporting them should be our priority. 

> I still think it may not be worth it to implement this feature at the moment. It perhaps could be postponed until `Alpha v0.1.0`.

I recommend postponing this feature until a more appropriate release date, `v1.0.0`? I think that implementing it now would be very time consuming and somewhat pointless.

**Comment by jackiedorland** [comment]
What user-agent does haste-client send out? If we could detect and parse it, we could potentially just make it work by redirecting requests with that useragent to /api/v0 instead.

**Comment by jackiedorland** [comment]
And, yeah. This could definitely be a v1.0.0 feature; I think UA parsing might be the way to go here, I really hope `haste` sends one out. Let me comb through the source for it:

**Comment by jackiedorland** [comment]
Damn. `haste-client` doesn't emit a User-Agent or any identifying information about the client. The only way we could know is through 
```
{
	"data": "...",
	"key: "..."
}
```

**Comment by lukewhrit** [comment]
> Damn. `haste-client` doesn't emit a User-Agent or any identifying information about the client. The only way we could know is through
> 
> ```
> {
> 	"data": "...",
> 	"key: "..."
> }
> ```

I'm almost sure `haste-client` sends out a user agent somewhere, it's probably just done automatically by [Faraday](https://github.com/lostisland/faraday).

I could also just fork `haste-client` and make modifications so that it works with spacebin.

**Comment by lukewhrit** [comment]
> Damn. `haste-client` doesn't emit a User-Agent or any identifying information about the client. The only way we could know is through
> 
> ```
> {
> 	"data": "...",
> 	"key: "..."
> }
> ```

`haste-client` uses Faraday as its HTTP request library, and Faraday does emit a User-Agent, Containing the word "Faraday" and it's current version. This means `haste-client` emits the following User-Agent: `Faraday v0.9`.

Although `haste-client` does emit a user agent I think it would be better to fork `haste-client` and modify it to make it work.

**Comment by lukewhrit** [comment]
I don't think we should do this anymore. It's not something I see being highly used, and our API is quickly becoming way different than haste's.
 
Implementing this would make our API a lot messier, harder to maintain, and probably cause some confusion.

I'm going to close this issue, If we do decide to implement this we can reopen.
